OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/ImportKeysFileFragment.java

```diff
@@ -104,20 +124,133 @@ public void onAttach(Activity activity) {
     @Override
     public void onActivityResult(int requestCode, int resultCode, Intent data) {
         switch (requestCode) {
-            case REQUEST_CODE_FILE: {
+            case REQUEST_CODE_FILE:
                 if (resultCode == Activity.RESULT_OK && data != null && data.getData() != null) {
+                    mCurrentUri = data.getData();
 
-                    // load data
-                    mImportActivity.loadCallback(new ImportKeysListFragment.BytesLoaderState(null, data.getData()));
+                    if (checkAndRequestReadPermission(mCurrentUri)) {
+                        startImportingKeys();
+                    }
                 }
-
                 break;
-            }
+            case REQUEST_CODE_IMPORT_ENCRYPTED_KEY:
+                if (resultCode == Activity.RESULT_OK) {
+                    InputDataResult result = data.getParcelableExtra(InputDataResult.EXTRA_RESULT);
+
+                    if (result.getResult() == InputDataResult.RESULT_ERROR) {
+                        result.createNotify(mImportActivity).show();
+                        return;
+                    }
+
+                    ArrayList<Uri> outputUris = result.getOutputUris();
+                    if (outputUris.size() != 1) {
+                        Notify.create(mImportActivity, R.string.error_bad_data, Style.ERROR).show();
+                        return;
+                    }
 
+                    mImportActivity.loadCallback(new BytesLoaderState(null, outputUris.get(0)));
+                }
+                break;
             default:
                 super.onActivityResult(requestCode, resultCode, data);
+        }
+    }
 
-                break;
+    private void startImportingKeys() {
+        boolean isEncrypted;
+        try {
+            isEncrypted = isEncryptedFile(mCurrentUri);
+        } catch (IOException e) {
+            e.printStackTrace();
+            Notify.create(mImportActivity, R.string.error_bad_data, Style.ERROR).show();
+            return;
+        }
+
+        if (isEncrypted) {
+            Intent intent = new Intent(getActivity(), DecryptActivity.class);
+            intent.setAction(DecryptActivity.ACTION_DECRYPT_FILE_AND_RETURN);
```

Comment:
Yes. That was my first try, but i spent a lot of time on it with not very good results regarding to user experience.
I thought the roundtrip was better for some reasons.

But i've just had some new ideas, i will try them and report back.
Thank you for the review.